mtags-shared/src/main/scala/scala/meta/internal/metals/PackageIndex.scala

```diff
@@ -88,7 +84,56 @@ class PackageIndex() {
     )
   }
 
-  private def visitJarEntry(jarpath: AbsolutePath): Unit = {
+  /**
+   * Returns the parent directory of the absolute string path
+   *
+   * Examples:
+   *
+   * {{{
+   *   dirname("/a/b/") == "/a/"
+   *   dirname("/a/b") == "/a/"
+   *   dirname("/") == "/"
+   * }}}
+   *
+   * @param abspath
+   *   a string path that matches the syntax of ZipFile entries.
+   */
+  def dirname(abspath: String): String = {
+    val isDir = abspath.endsWith("/")
+    val end =
+      if (isDir) abspath.lastIndexOf('/', abspath.length - 2)
+      else abspath.lastIndexOf('/')
+    if (end < 0) "/"
+    else abspath.substring(0, end + 1)
+  }
+
+  /**
+   * Returns the name of top-level file or directory of the absolute string path
+   *
+   * Examples:
+   *
+   * {{{
+   *   basename("/a/b/") == "b"
+   *   basename("/a/b") == "b"
+   *   basename("/") == ""
+   * }}}
+   *
+   * @param abspath
+   *   a string path that matches the syntax of ZipFile entries.
+   */
+  def basename(abspath: String): String = {
```

Comment:
Are these methods copied over from PathIO ?

Reply:
Yes. I noticed that after introducing `mtags-shared` I had to replace all `AbsolutePath` and `RelativePath` usages to NIO. It led to a situation like this, where some code was handled in scalameta parser and forced me to copy it. It also introduced many inconsistencies in the code. Maybe it is worth addressing the whole issue by doing some refactor / unification of those paths in the code ?

Reply:
This is fine, no worries.